servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java

```diff
@@ -231,33 +232,41 @@ public boolean isAvailable() {
             return retainedAddresses.isEmpty() ? null : retainedAddresses.values();
         }
 
-        @Nullable
-        Collection<E> consume(final Collection<E> events) {
+        Collection<E> consumeAndFilter(final Collection<E> events) {
             if (retainedAddresses == NONE_RETAINED) {
+                List<E> toReturn = new ArrayList<>(events.size());
                 for (E e : events) {
                     if (e.isAvailable()) {
-                        activeAddresses.put(e.address(), e);
+                        if (activeAddresses.putIfAbsent(e.address(), e) == null) {
+                            // FIXME: what if it's a custom E that has more fields?
+                            toReturn.add(e);    // filter out duplicated addresses
+                        }
                     } else {
-                        activeAddresses.remove(e.address());
+                        if (activeAddresses.remove(e.address()) != null) {
+                            toReturn.add(e);   // filter out addresses that were already unavailable
+                        }
                     }
+                    // FIXME: remove addresses that are available and unavailable in the same events collection
```

Comment:
If `Collection` has two entries:
1. foo, available
2. foo, unavailable
we will propagate both events to RRLB instead of propagating nothing. It will add more work for RRLB to add the address and remove it immediately. Due to concurrency between threads, that address can be used by a client before the 2nd event is processed